### ISSUE TITLE
Test out Xarray Tensorstore Backend

### ISSUE THREAD

**Issue by jacobbieker** [openclimatefix/ocf_datapipes#198]
https://github.com/google/xarray-tensorstore

## Detailed Description
We have been trying to speed up access from zarr for quite awhile. Tensorstore might help, and Google recently made public a backend for xarray that uses Tensorstore.

@JackKelly

**Comment by assafshouval** [comment]
I would like to take this issue. I'm new here. how these stuff works?
Where to start look at?

**Comment by jacobbieker** [comment]
Hi! That's great! For this, the primary place that things would need to be updated would be the files in `ocf_datapipes/load`, as this issue is primarily concerned with opening up and reading the Zarrs with tensorstore. The two main ones you would want to look at would be https://github.com/openclimatefix/ocf_datapipes/blob/087149831f785903b600ae10540895ff3b6e511c/ocf_datapipes/load/satellite.py#L56 for satellite data, which is available on GCP [here](https://console.cloud.google.com/marketplace/product/bigquery-public-data/eumetsat-seviri-rss-hrv-uk?project=tactile-acrobat-249716) if you want some more to test with, other than the unit tests which include a bit of satellite data. And for NWP data sources, primarily the ICON data: https://github.com/openclimatefix/ocf_datapipes/blob/087149831f785903b600ae10540895ff3b6e511c/ocf_datapipes/load/nwp/providers/icon.py#L8 We have an archive of ICON data in Zarr format on HuggingFace [here](https://huggingface.co/datasets/openclimatefix/dwd-icon-eu) where you can download an example or two to try things out. Although there is also some data included in this repo for the unit tests, incase you want to try with that instead too. 

Ideally, it should be possible to just swap out the `xr.open_zarr` with `xarray_tensorstore.open_zarr` with minimal changes. There are some caveats though:
1. From some initial testing, TensorStore possibly doesn't work very well with our compression algorithm `ocf_blosc2` @dfulu might have more info
2. We don't want to use the `xarray-tensorstore.read()` until we have cropped and picked out examples to use, otherwise it seems like it might read into memory the whole dataset, which for most of our data sources is multiple TBs.

**Comment by assafshouval** [comment]
@jacobbieker, just updating that I've tried to switch to xarray_tensorstore.open_zarr(path), but I've got the following problem in the line after when sortby('time').
`dataset = dataset.drop_duplicates("time").sortby("time")`
I've encountered the same problem and opened an issue here: [https://github.com/google/xarray-tensorstore/issues/1#issue-1855187534](url)
I'm wondering whether I don't have the dependencies right...

**Comment by assafshouval** [comment]
> > @jacobbieker, just updating that I've tried to switch to xarray_tensorstore.open_zarr(path), but I've got the following problem in the line after when sortby('time'). `dataset = dataset.drop_duplicates("time").sortby("time")` I've encountered the same problem and opened an issue here: [https://github.com/google/xarray-tensorstore/issues/1#issue-1855187534](https://github.com/google/xarray-tensorstore/issues/1) I'm wondering whether I don't have the dependencies right...
> 
> I thought xarray-tensorstore uses Zarr-python for the metadata, so the sorting should work with that, but yeah, sorry its not much help.

Yeah, they do use zarr-python, but the problem is when trying to deep copy the dataset. 
I'll invest a little more time exploring this, and see if I can advance this further, if not I'll leave it for now.
Thanks

**Comment by shoyer** [comment]
> @jacobbieker, just updating that I've tried to switch to xarray_tensorstore.open_zarr(path), but I've got the following problem in the line after when sortby('time'). `dataset = dataset.drop_duplicates("time").sortby("time")` I've encountered the same problem and opened an issue here: [https://github.com/google/xarray-tensorstore/issues/1#issue-1855187534](url) I'm wondering whether I don't have the dependencies right...

This should be fixed in the latest 0.1.1 release of Xarray-Tensorstore.

**Comment by assafshouval** [comment]
indeed, it works. thanks @shoyer.
@jacobbieker I have more questions:
a. Do you have a good example that would beneficial to bench-mark it?
b. It doesn't support opening multi-file dataset as in:
dataset = xr.open_mfdataset(zarr_path, **openmf_kwargs), for example in sattelite.py, line 45. I don't how much this scenario is worth investing time in, but first maybe worth to benchmark the first case ...

**Comment by jacobbieker** [comment]
Hi, yeah, a good benchmark would be a single satellite zarr, like this one: gs://public-datasets-eumetsat-solar-forecasting/satellite/EUMETSAT/SEVIRI_RSS/v4/2023_hrv.zarr

And okay, thanks, yeah, if it is a lot faster, we can probably fine a workaround to that issue.

**Comment by jacobbieker** [comment]
Update from my testing: Tensorstore does not support compressors not on this list https://google.github.io/tensorstore/driver/zarr/index.html#json-driver/zarr/Compressor and so can't open most of the OCF Zarrs which are compressed with `Blosc2`
